samza-core/src/main/java/org/apache/samza/execution/LocalJobPlanner.java

```diff
@@ -119,21 +137,79 @@ private void createStreams(String planId, List<StreamSpec> intStreams, StreamMan
+      MetadataStore metadataStore = getMetadataStore();
+      DistributedLock distributedLock = coordinationUtils.getLock(lockId);
+      if (distributedLock == null || metadataStore == null) {
+        LOG.warn("Processor {} failed to create utils. Each processor will attempt to create streams.", uid);
+        // each application process will try creating the streams, which requires stream creation to be idempotent
         streamManager.createStreams(intStreams);
-        lockWithState.unlockAndSet();
-      } else {
-        LOG.info("Processor {} did not obtain the lock for streams creation. They must've been created by another processor.", uid);
+        return;
       }
-    } catch (TimeoutException e) {
-      String msg = String.format("Processor {} failed to get the lock for stream initialization", uid);
-      throw new SamzaException(msg, e);
+      //Start timer for timeout
+      long startTime = System.currentTimeMillis();
+      long lockTimeout = TimeUnit.MILLISECONDS.convert(CoordinationConstants.LOCK_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+
+      // If "stream created state" exists in store then skip stream creation
+      // Else acquire lock, create streams, set state in store and unlock
+      // Checking for state before acquiring lock to prevent all processors from acquiring lock
+      // In a while loop so that if two processors check state simultaneously then
+      // to make sure the processor not acquiring the lock
+      // does not die of timeout exception and comes back and checks for state and proceeds
+      while ((System.currentTimeMillis() - startTime) < lockTimeout) {
+        if (metadataStore.get(String.format(STREAM_CREATED_STATE_KEY, lockId)) != null) {
+          LOG.info("Processor {} found streams created state data. They must've been created by another processor.", uid);
+          break;
+        }
+        try {
+          if (distributedLock.lock(10000, TimeUnit.MILLISECONDS)) {
+            LOG.info("lock acquired for streams creation by Processor " + uid);
+            streamManager.createStreams(intStreams);
+            String streamCreatedMessage = "Streams created by processor " + uid;
+            metadataStore.put(String.format(STREAM_CREATED_STATE_KEY, lockId), streamCreatedMessage.getBytes("UTF-8"));
+            break;
+          } else {
+            LOG.info(
+                "Processor {} did not obtain the lock for streams creation. They must've been created by another processor.",
+                uid);
+          }
+        } catch (TimeoutException e) {
+          LOG.warn("Processor {} failed to get the lock for stream initialization. Will try again until time out", uid);
+        } catch (UnsupportedEncodingException e) {
+          String msg = String.format("Processor {} failed to encode string for stream initialization", uid);
+          throw new SamzaException(msg, e);
+        } finally {
+          distributedLock.unlock();
```

Comment:
It seems possible for the processor to not obtain `lock` but still call `unlock`. Can that potentially have any side effects?